### PR TITLE
monorepo vs polyrepo

### LINKED ISSUE #65: monorepo vs polyrepo
Monorepo:
- pros and cons

Polyrepo:
- pros and cons

### PR DESCRIPTION
**Pull request by olivierschultz**
closes #65

### CODE REVIEW

File: software-development-strategy/monorepo_vs_polyrepo.md [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    </tr>
+    <tr>
+      <th>Changes</th>
+      <td>Ensures changes affect all the projects, can be tracked together, tested together, and released together.</td>
+      <td>Ensures changes affect only one project, can be tracked separately, tested separately, and released separately.</td>
+    </tr>
+    <tr>
+      <th>Collaboration</th>
+      <td>Encourages collaboration and code sharing within an organization. </td>
+      <td>Encourages collaboratio and code sharing across organizations.</td>
```
Comment (BerthelotAnthony) [inline software-development-strategy/monorepo_vs_polyrepo.md]: collaboratio**n**

File: software-development-strategy/monorepo_vs_polyrepo.md [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    <tr>
+      <th>Thinking</th>
+      <td>Encourages thinking about conjoins among projects.</td>
+      <td>Encourages thinking about contracts between projects.</td>
+    </tr>
+    <tr>
+      <th>Access</th>
+      <td><p>Access control defaults to all projects.<p>Some teams use tools for finer-grained access control.<p>Gitlab offers ownership control where you can say who owns what directories for things like approving merge requests that affect those directories. Google Piper has finer-grained access control. Phabricator offers herald rules that stop a merge from happening if a file has changed in a specific subdirrectory. Some teams use service owners, so when a change spans multiple services, they are all added automatically as blocking reviewers.</td>
+      </td>
+      <td><p>Access control defaults to per project.<p>Some teams use tools for broader-graned access control.<p>GitHub offers teams where you can say one team owns many projects and for things like approving requests that affect multiple repos.</td>
```
Comment (BerthelotAnthony) [inline software-development-strategy/monorepo_vs_polyrepo.md]: defaults per project ?

File: software-development-strategy/monorepo_vs_polyrepo.md [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    </tr>
+  </tbody>
+</table>
+
+## Monorepo scaling
+
+Monorepo scaling becomes a problem when a typical developer can't work well with the code by using typical tools such as vanilla git.
+
+* Monorepo scaling eventually becomes impractical in terms of space.
+* Monorepo scaling eventually becomes impratical in terms of time.
```
Comment (BerthelotAnthony) [inline software-development-strategy/monorepo_vs_polyrepo.md]: impra**c**tical

File: software-development-strategy/monorepo_vs_polyrepo.md [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+* Monorepo may grow so large contain, so many projects that it takes too much mental effort to work accross projects.
+
+Monorepo scaling can be improved by:
+* Some type of virtual file system (VFS) that allows a portion of the code to be present locally.
+* Sophisticated source code indexing/searching/discovery capabilities as a service.
+
+Monorepo scaling seems to become an issue, in pratice, at approximately these kinds of metrics:
+* 10 - 100 developers writing code full time.
+* 10 - 100 projects in progress at the same time.
+* 10 - 100 packaging process during the same time period.
```
Comment (BerthelotAnthony) [inline software-development-strategy/monorepo_vs_polyrepo.md]: process**es**

File: software-development-strategy/monorepo_vs_polyrepo.md [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+* 1K - 10K lines of code
+* 1K - 10K versioned dependencies
+
+Proponents of monorepo:
+* If components need to release together, then use a monorepo.
+* If components need to share common code, then use a monorepo.
+* In less-mature, high-churn codebase, use a monorepo.
+
+Advantages of monorepo:
+* Ease of code reuse - similar funtionality or communication protocols can be abstracted into shared libraries and directly included by projects, without the need of a dependency package manager
```
Comment (BerthelotAnthony) [inline software-development-strategy/monorepo_vs_polyrepo.md]: functionality

File: software-development-strategy/monorepo_vs_polyrepo.md [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+* Simplified dependency management - In a multiple repository environment where multiple projects depend on a thrid-party dependency, that dependency might be downloaded or built multiple times.
```
Comment (BerthelotAnthony) [inline software-development-strategy/monorepo_vs_polyrepo.md]: th**ir**d party

**Comment by olivierschultz** [comment]
@BerthelotAnthony I think I'm done with this PR, can you check it out pls?

File: software-development-strategy/monorepo_vs_polyrepo.md [hunk 7]
```diff
@@ -0,0 +1,228 @@
+# Software Development Strategy - Monorepo vs. Polyrepo
+
+## Introduction
+
+### What is monorepo?
+
+Monorepo (syllabic abbreviation of monolithic repository) is a nickname that means "using one repository for the source code management version constrol system (VCS)".
```
Comment (BerthelotAnthony) [inline software-development-strategy/monorepo_vs_polyrepo.md]: **control**

File: software-development-strategy/monorepo_vs_polyrepo.md [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+  - git shallow clone
+  - git filter-branch
+  - git clone only one branch
+
+### What is polyrepo?
+
+Polyrepo (syllabic abbreviation of poly repository) is a nickname that means "using multiple repositories for the source code management version control system (VCS)".
+
+Advantages of polyrepo:
+* Strict isolsation of compenents
```
Comment (BerthelotAnthony) [inline software-development-strategy/monorepo_vs_polyrepo.md]: comp**o**nents

Comment (BerthelotAnthony) [inline software-development-strategy/monorepo_vs_polyrepo.md]: **isolation**

**Pull request merged**